### ISSUE TITLE
Centering functionality

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#90]
@PyAbel/developers, @ldes89150 

I was trying to incorporate image centering into the abel.iabel/fabel functions and it made me realize that our centering functions are fairly old and probably not ideal in light of what we have learned recently.

So, we should ask the question about what kind of centering functions that we would like to have. I am especially dissatisfied with my old "tools.symmetry.center_image" function, since it forces us to specify the desired image size (`n`) instead of just cropping the image to the edge of the data. So, first of all, I think that it's important to have a `center_image` function that allows the user to specify the center of the image and then crops the image to the correct size.

In addition, @stggh has shown that it's fairly trivial to use [scipy.ndimage.interpolation.shift](http://docs.scipy.org/doc/scipy-0.16.1/reference/generated/scipy.ndimage.interpolation.shift.html#scipy.ndimage.interpolation.shift) in order to specify an arbitrary – subpixel – center. In addition, an odd-pixel-width image can always be obtained. (@stggh, correct me if I'm wrong here.) Fundamentally, this process involves some interpolation, so it is more computationally expensive, and in principal, it could blur pixel-scale features (though I think that in real-world situations, it's basically equivalent to shifting your CCD camera my 0.5 pixels, which should not matter). 

So, we also need to make some decision of how we want to support integer pixel centers versus sub-pixel (float) centers. 

Finally, we need to decide if we want functions that find the center (like the new, `tools.center.find_center` function), center an image (`tools.symmatry.center_image`), or functions that do both (like @stggh's `tools.vmi.find_image_center_by_slice`). 

I would recommend that we keep finding the center and setting the center separate in order to minimize code repetition and confusion (setting the image center is always super confusing!).

Also, let me check with @DhrubajyotiDas: BASEX no longer requires square images, correct? If so, we should move away from centering functions that aim to output square images.

So, here is what I propose:

1) We re-write `tools.symmetry.center_image` to allow it to automatically crop the image to the correct size given the region of data validity. We can still keep the functionality for the user to specify n_horiz and n_vert manually. This function should probably live in `tools.center.center_image`.

2) We include a keyword argument in our `center_image` function to allow it to re-interpolate sub-pixel images using scipy.ndimage.interpolation.shift. I don't think that it's a good idea to force every image to be re-interpolated, but we can have a keyword argument (`sub_pixel_center = False`) that allows the "user" to decide if the centering should be integer of sub-pixel. I put "user" in quotes, because in the abel.iabel/fabel function, we will basically make this decision for the user. If they specify an integer center, then we will use the integer method and not interpolate. If they select an automatic centering algorithm (of if they specify a float center) then we will user the interpolation method. 

3) We keep the "find_center" and "center_image" functions separate. If we want, we could provide a "find_center_and_center_image()" function. But, since we are going to combine these in the abel.iabel function, then I think that this should satisfy must users.

Any thoughts or volunteers to implement these changes?

**Comment by DhrubajyotiDas** [comment]
>Also, let me check with @DhrubajyotiDas: BASEX no longer requires square images, correct? If so, we should move away from centering functions that aim to output square images.

Correct.

**Comment by DanHickstein** [comment]
Great that is good to know. This makes the existing `center_image` function (which always generates a `nxn` image) seem very outdated.

**Comment by ldes89150** [comment]
Good point for using `scipy.ndimage.interpolation.shift `

How about something like this
```python
def center_image(data, center = 'com'):
    if isinstance(center, str):
        center = find_center(data, center)
    old_center  = data.shape[1]/2.0, data.shape[0]/2.0
    dx = center[0] - old_center[0]
    dy = center[1] - old_center[1]
    centered_data = shift(data, (dy,dx))
    return centered_data
```

**Comment by stggh** [comment]
@DanHickstein - good discussion 

1. `IModd = scipy.ndimage.interpolation.shift(IMeven, (-1/2, -1/2))[:-1, :-1]`

     will return a correctly centered odd image (for a correctly centered even image). There are of course other permutations, depending on what sides of the image are trimmed.

   BTW,
     (a) in principle only the image width (columns) needs to be odd.
     (b) quadrant analysis is only really needed where symmetry may be exploited
         `f/iabel_hansenlaw()` *et al.* are happy with 1/2 an image.

2. _`int` center_:  I cannot imagine how an experimental dataset may be perfectly aligned to a pixel grid to validate the use of an integer center. (Just as it is incorrect to use integer pixel intensity values). Sub-pixel centering is essential.  `f/iabel()` do not care about pixels, all these functions require is an intensity vs `R`, with `R=0, ... , n` (well for hansenlaw at least). The important point is that `R=0` is the image center-line.

3. _Centering_: I believe the only choice should be to supply a centered image to `f/iabel`. The user has an option to use supplied centering algorithms. 

4. _Cropping_: `IM[row1:row2, col1:col2]` are you planning to wrap this slice in a function call?  Perhaps, this sort of interface may be better handled in the GUI, with red-lines identifying the crop-limits.

In summary,  I agree 2 separate functions, `find_image_center(IM, method="abc")` returning `float (axis=0, axis=1)` coordinates, and `center_image(IM, data_origin=(axis=0, axis=1))` which will place the image at the image center_pixel.

@ldes89150  `get_image_quadrants()` does no more than split an image into quadrants. At the moment the quadrant combining is processed in the `AbelTransform` class instance function. In the future these processes may be moved to separate class methods, which might incorporate Fourier processing etc.

**Comment by DanHickstein** [comment]
@ldes89150, To give users the most flexibility, perhaps we should have three functions: `find_center()`, `set_center()` and `center_image()`, where center image is basically just `set_center( find_center(image))`.

Also, don't we need to make a decision about what data to keep after we use the scipy.ndimage.interpolation.shift? Basically, how should `set_center()` operate. For example:

```python
In [14]: Z = np.ones((5,5)); Z[1,1] = 2; print Z
[[ 1.  1.  1.  1.  1.]
 [ 1.  2.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]]

In [15]: scipy.ndimage.interpolation.shift(Z,(1,1))
Out[15]:
array([[ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  1.,  1.,  1.,  1.],
       [ 0.,  1.,  2.,  1.,  1.],
       [ 0.,  1.,  1.,  1.,  1.],
       [ 0.,  1.,  1.,  1.,  1.]])
```

In this situation, zeros have been padded along the top and left and data has been lost along the right. Is this the behavior that we want? It seems like there would be three options:

1) Size maintaining – What just happened, where data was removed and zeros were added.
2) Data maintaining – Padding zeros to make in the above case, a 7x7 image.
3) Symmetry maintaining – Only removing data, to make, in the above case, a 3x3 image.

Some users may have images that are far from center, so the difference between these could be huge. And I could see users wanting any of these options depending on the situation. Should we give them the option of these different behaviors through a keyword argument?

@stggh, yes, you are correct that for __experimental__ data the pixels are meaningless and it is probably fine to interpolate. But, what if a user has __simulated__ data with pixel-sized objects? The abel transform is generic and not only for experimentally obtained arrays. I would think that we should provide an option to specify an integer pixel-center where no interpolation is applied. 

I can also easily imagine situations where the user has a large image and only the center is the interesting part. I would suggest that we add an optional "crop" keyword to the center_image function that allows the user to crop the data to a certain size (i.e., crop=(500,700), where crop defaults to ('none','none') or something).

**Comment by DhrubajyotiDas** [comment]
> It seems like there would be three options:

@DanHickstein There's also a compromise option, where the average of the "remnant" pixels could be spread around the edges of the centered image to preserve the image size, while maintaining most of the data and symmetry. I was also struggling with this concept in my up/down symmetry function (during the process of removing the symmetric BASEX routine). Currently my up/down symmetry function simply crops out the extraneous data, but I realize that's not always the ideal case.

**Comment by ldes89150** [comment]
@DanHickstein Good point for `crop` keyword. That is exactly what I want!!

**Comment by stggh** [comment]
@DanHickstein  [`scipy.ndimage.interpolation.shift`](http://docs.scipy.org/doc/scipy-0.16.1/reference/generated/scipy.ndimage.interpolation.shift.html) appears to have many options covered.
```python
mode : str, optional

    Points outside the boundaries of the input are filled according to the given mode (‘constant’, ‘nearest’, ‘reflect’ or ‘wrap’). Default is ‘constant’.
```

I am a little concerned about our propensity to wrap standard functions, when say `shift(IM[row1:row2, col1:col2], shift options)`, might do. There are probably clever ways to unpack tuples, something like `shift(IM[*crop], (dy,  dx))` using some [smart tuple unpacking] (http://stackoverflow.com/questions/2238355/what-is-the-pythonic-way-to-unpack-tuples). However, I do understand, the need to provide convenient functions for users, so not a big deal, and not a show stopper.

**Comment by DanHickstein** [comment]
@DhrubajyotiDas, That sounds interesting, but I don't quite understand what you mean by spreading the "average of the remnant pixels around the edges".

@stggh, the different modes for `scipy.ndimage.interpolation.shift` all produce the same size array (5x5 in the case of my example above). You can basically just control what numbers are filled in the region outside of the image where `shift` by default fills zeros. Zeros seem fine to me. 

But, for my data, I would tend to prefer to run in Mode #3 (producing a 3x3 image) since each pixel value corresponds to actual data and there is no need to fill in zeros.

By wrapping standard functions, are you saying that the `set_center` function would really just be wrapping the `shift(IM[row1:row2, col1:col2], shift options)`? There would be a few differences, mainly that the `set_center` function should not use shift if it is passed and integer center, and that the crop should work slightly differently, allowing the user to specify the final size of the image instead of the start and end rows/cols.

**Comment by DhrubajyotiDas** [comment]
@DanHickstein Apologies for not being clearer, it was a semi-coherent idea that I was playing around with (and doesn't seem to be any better than padding zeroes now that I think about it). 

Basically, if you imagine a vector where the last two elements are about to be cropped (so, the len(old vector) - len(new vector) = 2), my proposed option is to take the last two elements, average them, and add one to the top of new vector and one to the bottom, so len(old vector) = len(new vector).

Essentially, you are conserving the shape of data, and instead of adding zeroes, you are padding the vector with elements of roughly the same magnitude as noise already present in that vector.

**Comment by DanHickstein** [comment]
@DhrubajyotiDas, That makes sense. I think that the `method='nearest'` option for `shift` is basically trying to accomplish a similar thing by appending the nearest value instead of zeros. Might be a better option than zeros in some cases since it keeps the baseline/noise level constant.

**Comment by stggh** [comment]
@DanHickstein "wrapping", yes use the [scipy functions](http://docs.scipy.org/doc/scipy/reference/ndimage.html) directly if available, rather than renaming them. However, I see your point, about differenct functionality.

**Comment by DanHickstein** [comment]
I think that there are a few more improvements that need to be made to `center.py` before we can consider this centering issue closed:

1. Move the `abel.tools.vmi.find_image_center_by_slice` function to `tools.center.py` and make it work with the `find_center` function. @stggh seems like the one to do this.

2. Implement the other two cropping options in the `set_center` function. By default, scipy.ndimage.interpolate.shift() provides the `maintain_size` option. But, it's possible that users will also want the `maintain_symmetry` and `maintain_data` options described above.
A volunteer to implement this one? @DhrubajyotiDas, @ldes89150 @rth?

**Comment by ldes89150** [comment]
@DanHickstein I can help implement some `crop` features :D

**Comment by DanHickstein** [comment]
Awesome! Thanks @ldes89150! 

While you are at it, you could also add some docstrings to the centering functions.

**Comment by stggh** [comment]
1. Done see #98, PR #100 (century, yay)

**Comment by DanHickstein** [comment]
Thanks @stggh! 

(I'm not quite sure how the PR numbering works on GitHub. I think that we only have 50 PRs, so it must skip some numbers...)

**Comment by DhrubajyotiDas** [comment]
@DanHickstein I believe both PRs and Issues get sourced from the same pool of integers, and the last issue was 99.

**Comment by DanHickstein** [comment]
That make sense. So we are up to a century of Issues+PRs. That's pretty impressive :)

**Comment by ldes89150** [comment]
#101

**Comment by DanHickstein** [comment]
I'm happy with the way that the centering functions are working. Closing this. Good work team!
